### PR TITLE
AutoComplete on arguments, better CallTips, run multiple lines in shell, and see cpu and mem usage!

### PR DESCRIPTION
**Pull request by drorspei**
Hello!

I managed to make a pull request!! :)
But not more than one...

CallTips: I refactored a bit so it would be easier to implement the above functionality. Also, if the function has a tuple in its def or unpacking arguments, the correct names are now showed. For the former there is a new delist function, and for the latter it's a simple change in the loop. ALSO: the calltip now breaks long function defs into lines!!! YES!!!!

AutoComplete changes: if CallTips extension is enabled, it can be used to get argument names while typing a function call.

MultiLineRun: added right click menu to run a few selected lines. You select a few lines, right click, choose run lines, and they run.

ResourcesStatus: working on some big data I realised my ram keeps running out, and also I wanted to make sure that some parallel functions actually run in parallel (in external C libraries), so I added this.

Hope someone finds these useful.

Dror

### CODE REVIEW

**Comment by drorspei** [comment]
I should probably write how to test the changes...
Let's start with CallTips:
If you define the following function:

    def foo((a, b)=(1,2), *bar, **shimi):
        pass

then the old call tip would be:

    foo(<tuple>=(1, 2), *args, **kwargs)

and the new one is:

    foo((a, b)=(1, 2), *bar, **shimi)

**Comment by drorspei** [comment]
Checking AutoComplete change:
Define the following function

    def foo(bar): pass

Then start typing:

    foo(b

and click tab. The new option to complete with "bar=" should appear.
It also works correctly with the "eval" mechanism. Define:

    def foo():
        def bar(shimi): pass
        return bar

Then start typing

    foo()(sh

and hit tab. Nothing new should happen. Now press Ctrl-Space (key to force autocomplete), and the new option of "shimi=" should appear.

It doesn't complete unpacking arguments (*args and **kwards) names, which I think makes sense. It is very rare that I assign values to them instead of using the builtin unpacking mechanism.

File: idlesporklib/CallTips.py [hunk 1]
```diff
@@ -136,7 +151,20 @@ def _find_constructor(class_ob):
...
+        if isinstance(x, str):
+            items.append(x)
+        else:
+            items.append(delist(x))
+    if len(items) > 1:
+        return '({})'.format(', '.join(items))
+    else:
+        return '({},)'.format(items[0])
+
+def get_fob(ob):
```
Comment (idlespork) [inline idlesporklib/CallTips.py:173]: Can you change the documentation of this function to match your change?

File: idlesporklib/CallTips.py [hunk 2]
```diff
@@ -175,24 +203,58 @@ def get_arg_text(ob):
...
         fob = ob
+
+    return fob, arg_offset, argspec, ob_call
+
+def get_arg_names(ob):
+    fob, arg_offset, _, _ = get_fob(ob)
+    try:
+        argcount = fob.func_code.co_argcount
+        ret = list(fob.func_code.co_varnames[arg_offset:argcount])
+        ret = [arg for arg in ret if re.match("(?<!\d)\.\d+", arg) is None]
```
Comment (idlespork) [inline idlesporklib/CallTips.py]: Please add a comment why this is needed

File: idlesporklib/config-extensions.def [hunk 3]
```diff
@@ -54,6 +54,9 @@ force-open-calltip=<Control-Key-backslash>
 try-open-calltip=<KeyRelease-parenleft>
 refresh-calltip=<KeyRelease-parenright> <KeyRelease-0>
 
+[ResourcesStatus]
```
Comment (idlespork) [inline idlesporklib/config-extensions.def:59]: Either add psutil as a dependency (add install_requires=... to setup.py), or set the default to disabled. Nice extension!

**Comment by drorspei** [comment]
New feature: autocompletion on dict keys!
To test, type:

    {'hello':1, 'goodbye':2, None:3, (4,5):6}[

and press Ctrl-space (force autocomplete key binding). You now get the autocomplete window:

    'goodbye'
    'hello'
    (4, 5)
    None

Amazing!
Of course Ctrl-space means that ".keys()" is evaluated on the entity found before the "[" character.

One caveat is unicode strings. If you type:

    {1:2}[u'

and press Ctrl-space, then it uses file completion instead of key completion. This is because we would have to check if "u" is right before the ' or ", which is a bummer, and I never use unicode anyway, so... no.

File: idlesporklib/config-extensions.def (hunk 3 above)
Comment (drorspei) [reply idlesporklib/config-extensions.def:59]: ok disabled

File: idlesporklib/CallTips.py (hunk 1 above)
Comment (drorspei) [reply idlesporklib/CallTips.py:173]: done!

File: idlesporklib/CallTips.py (hunk 2 above)
Comment (drorspei) [reply idlesporklib/CallTips.py]: also done!

**Comment by drorspei** [comment]
yeyyyyyyyyyyyy!

**Comment by drorspei** [comment]
I'll be looking at the jedi module in the new ipython to steal stuff ^_^

**Comment by ofir-reich** [comment]
Can you make it so that pressing "Enter" auto-completes according to the current selection like pressing tab twice does, or pressing "(" or "[" ? Actually, while on that subject, can we do it so that pressing tab once (after opening the completion menu) would be enough, and you wouldn't have to press twice?

And good work on all of this!

Thanks!

**Comment by drorspei** [comment]
yeah! that's a great idea.
made a new pull request with this :)

**Pull request merged**